sklearn/mixture/base.py

```diff
@@ -0,0 +1,425 @@
+            raise ValueError("Invalid value for 'n_init': %d "
+                             "Estimation requires at least one run"
+                             % self.n_init)
+
+        if self.max_iter < 1:
+            raise ValueError("Invalid value for 'max_iter': %d "
+                             "Estimation requires at least one iteration"
+                             % self.max_iter)
+
+        if self.reg_covar < 0:
+            raise ValueError("Invalid value for 'reg_covar': %.5f "
+                             "regularization on covariance must be "
+                             "non-negative"
+                             % self.reg_covar)
+
+        # Check all the parameters values of the derived class
+        self._check_parameters()
+
+    @abstractmethod
+    def _check_parameters(self):
+        """Check initial parameters of the derived class."""
+        pass
+
+    def _initialize_parameters(self, X):
+        """Initialize the model parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+        """
+        n_samples = X.shape[0]
+        random_state = check_random_state(self.random_state)
+
+        if self.init_params == 'kmeans':
+            resp = np.zeros((n_samples, self.n_components))
+            label = cluster.KMeans(n_clusters=self.n_components, n_init=1,
+                                   random_state=random_state).fit(X).labels_
+            resp[range(n_samples), label] = 1
+        elif self.init_params == 'random':
+            resp = random_state.rand(X.shape[0], self.n_components)
+            resp = resp / resp.sum(axis=1)[:, np.newaxis]
+        else:
+            raise ValueError("Unimplemented initialization method '%s'"
+                             % self.init_params)
+
+        self._initialize(X, resp)
+
+    @abstractmethod
+    def _initialize(self, X, resp):
+        """Initialize the model parameters of the derived class.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        resp : array-like, shape = (n_samples, n_components)
+        """
+        pass
+
+    def fit(self, X, y=None):
+        """Estimate model parameters with the EM algorithm.
+
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `max_iter`
+        times until the change of likelihood or lower bound is sless than
+        `tol`, otherwise, a `ConvergenceWarning` is raised.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self
+        """
+        X = _check_X(X, self.n_components)
+        self.n_features = X.shape[1]
```

Comment:
If you really need to store a data dependent value on the estimator object during fit, it should either not be public or be marked as a fitted parameter by using a `_` suffix to make it explicit that this is not an hyperparameter of the model. hyperparameter are only set via the constructor and the `set_params` method. See: http://scikit-learn.org/stable/developers/contributing.html#apis-of-scikit-learn-objects